src/OrchardCore/OrchardCore.ReCaptcha.Core/TagHelpers/ReCaptchaTagHelper.cs

```diff
@@ -22,16 +19,22 @@ namespace OrchardCore.ReCaptcha.TagHelpers
     public class ReCaptchaTagHelper : TagHelper
     {
         private readonly IResourceManager _resourceManager;
-        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly IEnumerable<IDetectRobots> _robotDetectors;
         private readonly ReCaptchaSettings _settings;
         private readonly ILogger _logger;
         private readonly ILocalizationService _localizationService;
-        private readonly IStringLocalizer S;
+        private readonly IStringLocalizer<ReCaptchaTagHelper> S;
 
-        public ReCaptchaTagHelper(IOptions<ReCaptchaSettings> optionsAccessor, IResourceManager resourceManager, ILocalizationService localizationService, IHttpContextAccessor httpContextAccessor, ILogger<ReCaptchaTagHelper> logger, IStringLocalizer<ReCaptchaTagHelper> localizer)
+        public ReCaptchaTagHelper(
+            IOptions<ReCaptchaSettings> optionsAccessor,
+            IResourceManager resourceManager,
+            ILocalizationService localizationService,
+            IEnumerable<IDetectRobots> robotDetectors,
```

Comment:
I am not sure you can resolve services like this. If it was doing it in the main method it's probably because it can vary on each execution and the the tag is a singleton. @jtkech ?

Reply:
> IEnumerable<IDetectRobots> robotDetectors,

I did simlar thing in the PR #8408 but @deanmarcussen don't advice using this due the transient registration

Reply:
@sebastienros @Diman96 @hishamco 

TagHelpers are activated by a singleton factory but in a scoped context

So, normally nothing prevents a tag helper from injecting `IDetectRobots` for which we provide an implementation registered as a transient service. **That said better to try it and see if it doesn't fails at runtime**

Note: Taghelpers can be registered as transients if you call `AddTagHelpersAsServices()` (but we don't), but just to retrieve the types to be activated in a scope context, as we do with our liquid activator but only with our taghelpers

Reply:
It's also transient, https://github.com/OrchardCMS/OrchardCore/blob/beea0ae534dd687bbdcc1a8c8f630bcb40c3633d/src/OrchardCore/OrchardCore.DisplayManagement/Extensions/TagHelperExtensions.cs#L23

Reply:
Yes, we register our tag helpers as transients, as mvc does when you call `AddTagHelpersAsServices()`, for liquid, it allows us to retrieve them through the DI which is faster, in place of activating them in a given view context by using the mvc tag helper factory.

So our tag helpers are transients but for liquid always resolved in a scoped context, as mvc does by retrieving the `RequestServices` from the current view context. More precisely we resolve them from the current shell scope (see below), as we can have multiple isolated shell scopes (nested or not) in the same request scope, even more we can create a child scope whose parent container is related to another tenant.

See in `LiquidTagHelperActivator`
https://github.com/OrchardCMS/OrchardCore/blob/beea0ae534dd687bbdcc1a8c8f630bcb40c3633d/src/OrchardCore/OrchardCore.DisplayManagement.Liquid/TagHelpers/LiquidTagHelperActivator.cs#L126-L134

Reply:
> We tried tag helper works. But only with explicit import in viewimports

Yes for razor, and your module project file needs to reference the project / package that defines the tag helper, otherwise it would work on `Development` but not on `Production` where pre-compiled views (compiled in the context of the module) are used.

But not needed when used under liquid

> Do you mean that AddTagHelpers adds it as a singleton?

See my previous comment

For liquid, AddTagHelpers is not necssary but better, see my previous comment